File: app/src/main/java/com/github/borz7zy/telegramm/ui/SplashFragment.java
package com.github.borz7zy.telegramm.ui;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LiveData;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.github.borz7zy.telegramm.AppManager;
import com.github.borz7zy.telegramm.R;
import com.github.borz7zy.telegramm.core.accounts.AccountEntity;
import com.github.borz7zy.telegramm.core.accounts.AccountManager;
import com.github.borz7zy.telegramm.core.accounts.AccountSession;
import com.github.borz7zy.telegramm.core.accounts.AccountStorage;
import com.github.borz7zy.telegramm.core.settings.SettingsEntity;

import org.drinkless.tdlib.TdApi;

import java.util.List;

public class SplashFragment extends Fragment {

    private TextView splashText;

    private boolean isAnimationDone = false;
    private TdApi.AuthorizationState pendingState = null;

    private LiveData<TdApi.AuthorizationState> currentAuthLiveData;

    private boolean navigated = false;

    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private Runnable blinkingRunnable;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_splash, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        splashText = view.findViewById(R.id.splashText);

        startTypewriterEffect();

        checkAndCreateFirstAccountIfNeeded();

        observeActiveAccount();
    }

    private void startTypewriterEffect() {
        String fullText = getString(R.string.app_name);
        splashText.setText("|");

        long totalDuration = 2000;
        long charDelay = totalDuration / fullText.length();

        new Runnable() {
            int c = 0;
            StringBuilder currentText = new StringBuilder();

            @Override
            public void run() {
                if (getContext() == null) return;

                if (c < fullText.length()) {
                    currentText.append(fullText.charAt(c));

                    splashText.setText(currentText.toString() + "|");

                    ++c;
                    mainHandler.postDelayed(this, charDelay);
                } else {
                    isAnimationDone = true;
                    startBlinkingCursor(fullText);

                    checkAndNavigate();
                }
            }
        }.run();
    }

    private void checkAndCreateFirstAccountIfNeeded() {
        AppManager.getInstance().getExecutorDb().execute(() -> {
            SettingsEntity settings = AppManager.getInstance()
                    .getAppDatabase()
                    .settingsDao()
                    .getSettings();

            if (settings != null && settings.currentActiveId != null) {
                // Уже есть активный аккаунт → ничего не делаем
                return;
            }

            // Проверяем, есть ли хоть какие-то аккаунты
            List<AccountEntity> accounts = AppManager.getInstance()
                    .getAppDatabase()
                    .accountDao()
                    .getAllAccounts();

            Integer activeId = null;

            if (!accounts.isEmpty()) {
                // Есть аккаунты, но не выбран активный → берём первый
                activeId = accounts.get(0).getAccountId();
            } else {
                // Совсем нет аккаунтов → создаём новый
                AccountEntity newAccount = new AccountEntity(null, 0L, "New User", "");
                AppManager.getInstance().getAppDatabase().accountDao().insert(newAccount);
                // После вставки accountId уже сгенерирован
                activeId = newAccount.getAccountId();
            }

            if (activeId != null) {
                AccountStorage.getInstance().setCurrentActive(activeId);
            }

            // После этого LiveData в AuthPhoneFragment и SplashFragment увидит аккаунт
        });
    }

    private void observeActiveAccount(){

        AccountStorage.getInstance()
                .observeActiveAccount()
                .observe(getViewLifecycleOwner(), account -> {

                    if(account == null){
                        pendingState = new TdApi.AuthorizationStateWaitPhoneNumber();
                        checkAndNavigate();
                        return;
                    }

                    AccountSession session =
                            AccountManager.getInstance()
                                    .getOrCreateSession(account);

                    observeSession(session);
                });
    }

    private void observeSession(AccountSession session){

        if(currentAuthLiveData != null){
            currentAuthLiveData.removeObservers(getViewLifecycleOwner());
        }

        currentAuthLiveData = session.observeAuthState();

        currentAuthLiveData.observe(getViewLifecycleOwner(), state -> {

            pendingState = state;
            checkAndNavigate();
        });
    }

    private void startBlinkingCursor(String text) {
        blinkingRunnable = new Runnable() {
            boolean showCursor = false;

            @Override
            public void run() {
                if (getContext() == null) return;

                if (showCursor) {
                    splashText.setText(text + "|");
                } else {
                    splashText.setText(text);
                }
                showCursor = !showCursor;

                mainHandler.postDelayed(this, 500);
            }
        };
        mainHandler.postDelayed(blinkingRunnable, 500);
    }

    private void checkAndNavigate(){
        if(navigated) return;
        if(isAnimationDone && pendingState != null){
            navigated = true;
            NavController nav =
                    Navigation.findNavController(requireView());
            if (pendingState instanceof TdApi.AuthorizationStateReady){
                nav.navigate(R.id.action_splashFragment_to_mainFragment);
            }else{
                nav.navigate(R.id.action_splashFragment_to_authPhoneFragment);
            }
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mainHandler.removeCallbacksAndMessages(null);
    }
}

File: app/src/main/java/com/github/borz7zy/telegramm/ui/SplashFragment.kt
package com.github.borz7zy.telegramm.ui

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.NavHostFragment
import com.github.borz7zy.telegramm.AppManager
import com.github.borz7zy.telegramm.R
import com.github.borz7zy.telegramm.core.accounts.AccountEntity
import com.github.borz7zy.telegramm.core.accounts.AccountManager
import com.github.borz7zy.telegramm.core.accounts.AccountStorage
import com.github.borz7zy.telegramm.ui.base.BaseTelegramFragment
import com.github.borz7zy.telegramm.utils.Logger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import org.drinkless.tdlib.TdApi
import org.drinkless.tdlib.TdApi.AuthorizationState
import org.drinkless.tdlib.TdApi.AuthorizationStateReady
import org.drinkless.tdlib.TdApi.AuthorizationStateWaitCode
import org.drinkless.tdlib.TdApi.AuthorizationStateWaitPassword
import org.drinkless.tdlib.TdApi.AuthorizationStateWaitPhoneNumber
import kotlin.coroutines.resume

class SplashFragment : BaseTelegramFragment() {

    private var splashText: TextView? = null
    private var typewriterJob: Job? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return inflater.inflate(R.layout.fragment_splash, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val root = view.findViewById<LinearLayout>(R.id.root_splash)
        splashText = view.findViewById(R.id.splashText)

        applyTheme(root)
        observeTheme(root)

        startTypewriterEffect()

        viewLifecycleOwner.lifecycleScope.launch {
            val state = waitForAuthState()
            handleNavigation(state)
        }
    }

    private fun applyTheme(root: LinearLayout?) {
        val theme = AppManager.getInstance().themeEngine.currentTheme.value
        root?.setBackgroundColor(theme?.surfaceColor ?: R.color.surfaceColor)
        splashText?.setTextColor(theme?.onSurfaceColor ?: R.color.onSurfaceColor)
    }

    private fun observeTheme(root: LinearLayout?) {
        AppManager.getInstance()
            .themeEngine
            .currentTheme
            .observe(viewLifecycleOwner) { theme ->
                root?.setBackgroundColor(theme.surfaceColor)
                splashText?.setTextColor(theme.onSurfaceColor)
            }
    }

    private suspend fun waitForAuthState(): TdApi.AuthorizationState {
        val account = getOrCreateAccount()

        val session = AccountManager.getInstance()
            .getOrCreateSession(account)

        return session
            .authStateFlow
            .filterNotNull()
            .first { state ->
                state !is TdApi.AuthorizationStateWaitTdlibParameters
            }
    }

    private suspend fun getOrCreateAccount(): AccountEntity =
        withContext(Dispatchers.IO) {

            suspendCancellableCoroutine { continuation ->

                AccountStorage.getInstance()
                    .getCurrentActive { account ->

                        if (account != null) {
                            continuation.resume(account)
                            return@getCurrentActive
                        }

                        val newAccount = AccountEntity(
                            null,
                            0L,
                            "New Account",
                            ""
                        )

                        val newId = AppManager.getInstance()
                            .appDatabase
                            .accountDao()
                            .insert(newAccount)

                        newAccount.setAccountId(newId.toInt())

                        AccountStorage.getInstance()
                            .setCurrentActive(newAccount.getAccountId())

                        Logger.LOGD(
                            "SplashFragment",
                            "Created new account ID: ${newAccount.getAccountId()}"
                        )

                        continuation.resume(newAccount)
                    }
            }
        }

    private fun handleNavigation(state: AuthorizationState?) {
        if (!isAdded) return

        val nav = NavHostFragment.findNavController(this)

        when (state) {
            is AuthorizationStateReady ->
                nav.navigate(R.id.frag_splash_to_main)

            is AuthorizationStateWaitPhoneNumber,
            is AuthorizationStateWaitCode,
            is AuthorizationStateWaitPassword ->
                nav.navigate(R.id.frag_splash_to_auth)

            else ->
                Logger.LOGD("SplashFragment", "Unknown auth state: $state")
        }
    }

    private fun startTypewriterEffect() {
        typewriterJob?.cancel()

        typewriterJob = viewLifecycleOwner.lifecycleScope.launch {

            val fullText = getString(R.string.app_name)
            val totalDuration = 2000L
            val charDelay = totalDuration / fullText.length

            splashText?.text = "|"

            val builder = StringBuilder()

            for (c in fullText) {
                builder.append(c)
                splashText?.text = "$builder|"
                delay(charDelay)
            }

            while (isActive) {
                splashText?.text = "$builder|"
                delay(500)
                splashText?.text = builder.toString()
                delay(500)
            }
        }
    }

    override fun onDestroyView() {
        typewriterJob?.cancel()
        super.onDestroyView()
    }
}
